### ISSUE TITLE
`KeyError: None` with dynamic hot-reload

### ISSUE THREAD

**Issue by oati** [pyedifice/pyedifice#97]
Trivial changes in the tutorial code causes a KeyError when using dynamic hot-reload. The application does not reload.

```
Traceback (most recent call last):
  File "/nix/store/yifccjf1gck1xz8cjhiihydy7z24wsjp-python3-3.11.6-env/lib/python3.11/site-packages/edifice/app.py", line 152, in event
    render_result = self._render_engine._refresh_by_class(classes)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/yifccjf1gck1xz8cjhiihydy7z24wsjp-python3-3.11.6-env/lib/python3.11/site-packages/edifice/engine.py", line 497, in _refresh_by_class
    if isinstance(self._component_tree[parent_comp], list):
                  ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
Stemming from these renders:
KeyError: None
```

I am using python 3.11.6, edifice 0.3.0, pyside6 6.6.0, qasync 0.27.0, and watchdog 3.0.0 on wayland (sway).

**Comment by jamesdbrock** [comment]
Hi @oati thanks for reporting this issue.

**Comment by jamesdbrock** [comment]
`git bisect` says that hot-reloading broke on this commit a4558f4f95a18f5b0ee7a9a20d2c661166e50139

**Comment by oati** [comment]
Hot-reloading is broken before https://github.com/pyedifice/pyedifice/commit/a4558f4f95a18f5b0ee7a9a20d2c661166e50139 with the same issue for me.

**Comment by jamesdbrock** [comment]
`git revert a4558f4f95a18f5b0ee7a9a20d2c661166e50139` does in fact restore the hot-reload.

**Comment by jamesdbrock** [comment]
Restoring just the `_attach_keys` function

```python
    def _render_base_component(self, component: BaseElement, render_context: _RenderContext) -> _WidgetTree:
        if len(component.children) > 1:
            self._attach_keys(component, render_context)
```

does not restore the hot-reload.

**Comment by jamesdbrock** [comment]
Note to self: part of what's happening is before commit a4558f4f95a18f5b0ee7a9a20d2c661166e50139 , the condition

```python
if (component.children[0].__class__ == old_children[0].__class__
```

would always be false, so then 

```python
self._update_old_component(old_children[0], component.children[0], render_context)
```

would never get called, and the new component would be `_render()`ed instead of updating the old component when we got to this line because it was never true that `child1 is not child2`:

https://github.com/pyedifice/pyedifice/blob/a4558f4f95a18f5b0ee7a9a20d2c661166e50139/edifice/engine.py#L634-L638

**Comment by jamesdbrock** [comment]
The question here is what does it mean for a class to be equal? We want different classes of 

https://github.com/pyedifice/pyedifice/blob/08d95633799556c08e7e7ab7996490eb4fcd1e67/edifice/_component.py#L667

to be different, which is why we compare by `__name__`:

https://github.com/pyedifice/pyedifice/blob/08d95633799556c08e7e7ab7996490eb4fcd1e67/edifice/engine.py#L365-L367

But also when a `ComponentElement` class is hot-reloaded, we want that to be a different class, maybe?

**Comment by jamesdbrock** [comment]
Hi @oati I just published a new version of Edifice which should hot-reload the tutorial correctly. Please tell me if you have any more issues.

**Comment by oati** [comment]
Update: It does work for the final tutorial code. It does not work for the example below.

```py
from edifice import App, Label, TextInput, View, Window, component

@component
def MyApp(self):
    meters_label_style = {"width": 170}
    feet_label_style = {"margin-left": 20, "width": 200}
    input_style = {"padding": 2, "width": 120}
    with Window():
        with View(layout="row", style={"margin": 10, "width": 560}):
            Label("Measurement in meters:", style=meters_label_style)
            TextInput("", style=input_style)
            Label("Measurement in feet:", style=feet_label_style)

if __name__ == "__main__":
    App(MyApp()).start()
```

**Comment by oati** [comment]
One possibility that would be consistent with my testing is that hot-reloading is currently broken for changes to the "top-level" or "root" component.

**Comment by oati** [comment]
Sure enough, this successfully updates:
```py
from edifice import App, Label, TextInput, View, Window, component

@component
def MyApp2(self):
    meters_label_style = {"width": 170}
    feet_label_style = {"margin-left": 20, "width": 200}
    input_style = {"padding": 2, "width": 120}
    with Window():
        with View(layout="row", style={"margin": 10, "width": 560}):
            Label("Measurement in meters:", style=meters_label_style)
            TextInput("", style=input_style)
            Label("Measurement in feet:", style=feet_label_style)

@component
def MyApp(self):
    MyApp2()

if __name__ == "__main__":
    App(MyApp()).start()
```

I did run into another potential bug (?): hot-reload in this case destroys and creates a new window unless both `Window` and `View` are in `MyApp`.

**Comment by jamesdbrock** [comment]
> I did run into another potential bug (?): hot-reload in this case destroys and creates a new window unless both `Window` and `View` are in `MyApp`, which is maybe intended behavior.

Yes that is a bug. It's hard to fix that bug because of the way that `Window` works. But the workaround, as you discovered, is adding another wrapping `View` child of the `Window`.

**Comment by jamesdbrock** [comment]
I just published v0.3.2 which preserves Hook state during hot-reloads.
